### PR TITLE
[cpu] add int8 sdpa api

### PR DESCRIPTION
**Pull request by Valentine233**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140108
* __->__ #138688

### Description:
- Add an independent sdpa api for int8 cpu path. Aim to only be used internally for pattern matcher. As discussed in https://github.com/pytorch/pytorch/pull/134317.
- Implement the math version for int8 sdpa.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -14778,6 +14778,11 @@
     CPU: _scaled_dot_product_flash_attention_cpu
   tags: nondeterministic_seeded
 
+- func: _scaled_dot_product_int8_for_cpu(Tensor query, Tensor key, Tensor value, float dropout_p=0.0, bool is_causal=False, *, Tensor? attn_mask=None, float? scale=None, int q_zp=0, float q_scale=1.0, int k_zp=0, float k_scale=1.0, int v_zp=0, float v_scale=1.0, int a_zp=0, float a_scale=1.0, int o_zp=0, float o_scale=1.0) -> Tensor output
```
Comment (jgong5) [inline aten/src/ATen/native/native_functions.yaml:14781]: Do we have to name it explicitly "for_cpu"? From the API semantics, it is not CPU specific?

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -5036,6 +5036,62 @@ def scaled_dot_product_flash_attention_for_cpu(
...
+    )
+    torch._check(
+        dropout_p == 0.0, lambda: f"dropout probability must be zero, got {dropout_p}"
+    )
+    torch._check(
+        query.shape[3] == value.shape[3] and key.shape[3] == value.shape[3],
+        lambda: "q, k, v should have the same head size",
+    )
+
+    output = aten._scaled_dot_product_int8_for_cpu.default(
```
Comment (jgong5) [inline torch/_decomp/decompositions.py:5077]: Suggest you add a note to link to that in `scaled_dot_product_flash_attention_for_cpu`, explaining why we need a decomposition that calls to the function itself (which sounds really tricky to me).

**Review by jgong5 — APPROVED** [review]

File: torch/_decomp/decompositions.py (hunk 2 above)
Comment (Valentine233) [reply torch/_decomp/decompositions.py]: Yes, you are right. The decomposition is removed.

**Comment by Valentine233** [comment]
Hi @drisspg, please help review the PR, thanks!

**Comment by Valentine233** [comment]
Hi @drisspg , please help review the PR when you are free, thanks! This is an essential PR for other int8 SDPA related ones.

**Comment by jgong5** [comment]
Hi @drisspg any comments on this PR? Thanks!

**Comment by Valentine233** [comment]
Hi @drisspg, thanks for your suggestions!

> Some immediate things that come to mind is we both have now had a scale parameter and a Q scale and a K scale, meaning we have three scales for the same matrix multiplication. Those scales would probably need to be on device if this was to work with accelerators. I'm `a_scale` is referring to the post-Softmax intermediate. That naming does not align with much of the literature.

In the new API, `a_scale` refers to the scale for attention and we can discuss more about the naming.

> Also, there's no fused kernel for this, so I don't understand why we even need this to be in C++. If the main goal is just to have an op for pattern matching, this could probably just be in Python.

The int8 kernel is fused, with parallelization on batch_size, num_heads and q blocks. We see **~20%** kernel and **~10%** E2E improvement for models like BertLarge, compared to the current bf16 flash attention.

> But if the main goal is pattern matching, I don't understand still because users will still need to make manual changes to call this new version of attention and then what is there to pattern match against?
> If there are limitations with how we do pattern matching or what is required for inductor, I think we should raise them and let's figure out better options.

Actually, when we enable the quantizations for MatMul, the graph is automatically suitable for int8 sdpa. The only thing we need to do is to add more int8 sdpa patterns. You could refer to the patterns 20-23 in https://github.com/pytorch/pytorch/pull/140108 as the current int8 patterns. I don't think any effort is needed for the user side.

> I'm still hesitant to put this in PyTorch Core because I think there are still questions on API signature and PyTorch Core has strong guarantees on FC/BC. I'll reiterate I think somewhere like TorchAO makes much more sense to place this where we are free to change things and allows us to iterate to find the best signature and eventually upstream the core. If there are reasons why this code can't live external to PyTorch Core, I think these are great data points and we should share them with the rest of the team to figure out if our ecosystem can support this type of extension.

To put int8 sdpa in TorchAO, we find a possible way: add the int8 kernel, pattern match, lowering in TorchAO, and then register a joint custom pass for int8 sdpa in PyTorch Core. Please give any suggestions if you have.

**Comment by Valentine233** [comment]
Decide to implement in TorchAO.

**Comment by Valentine233** [comment]
The int8 sdpa implementation in torchao has been explored, and we submitted a POC https://github.com/pytorch/ao/pull/1372. 
However, we find a big blocking issue: torchao has not the support for C++ AVX512/AVX2 compilation, which is needed for int8 sdpa kernel. It would take much efforts for the complete C++ building flag enabling. Do you think we could upstream the kernel in PyTorch Core like this PR, or should do the C++ building flag enabling in torchao? @malfet @drisspg

**Comment by jerryzh168** [comment]
I feel we should enable the building in torchao. but cc @msaroufim @atalman @jcaip as well if there is any comments as well

**Comment by jcaip** [comment]
@Valentine233 HI, can you share some more details on the C++ AVX512/AVX2 issues?

I agree with @jerryzh168 that enabling building with these flags in AO should be our long-term approach. TorchAO is meant to be the place for experimental performance features, if our build system is not set up to support those features, we should update, otherwise it will be difficult to convince people to contribute to AO.

**Comment by malfet** [comment]
@Valentine233 To add something like that to PyTorch core it needs a bit more discussion on how it will be used, what are the perf benefits and what are the accuracy tradeoffs and how hard to add something like that for CPU platforms other than x86? Or even, how this op will behave if CPU does not support AVX512?

As for enabling optional flags, this seems like a 4-5 line change in AO, I can help draft one if needed

**Comment by Valentine233** [comment]
Thanks for your comments! @jerryzh168 @jcaip @malfet

I would try to add the related building flag supports in TorchAO. Would ask for your help if encounter any issue.
